File: tests/lib/docker.ts
import { execSync } from "child_process";
import { resolve } from "path";

export class TestInfra {
  private composeFile: string;
  private projectName: string;

  constructor(caseName: string) {
    this.composeFile = resolve(import.meta.dirname, `../${caseName}/docker-compose.yml`);
    this.projectName = `chronos-e2e-${caseName}`;
  }

  start(): void {
    console.log(`[infra] Starting ${this.projectName}...`);
    execSync(
      `docker compose -f ${this.composeFile} -p ${this.projectName} up -d --wait`,
      { stdio: "inherit", timeout: 180_000 },
    );
    console.log(`[infra] ${this.projectName} is up`);
  }

  stop(): void {
    console.log(`[infra] Stopping ${this.projectName}...`);
    try {
      execSync(
        `docker compose -f ${this.composeFile} -p ${this.projectName} down -v --remove-orphans`,
        { stdio: "inherit", timeout: 60_000 },
      );
    } catch (e) {
      console.warn(`[infra] Warning during stop: ${e}`);
    }
  }
}

File: test/e2e/lib/docker.py
from __future__ import annotations

import subprocess
from pathlib import Path

E2E_DIR = Path(__file__).resolve().parents[1]


class TestInfra:
    def __init__(self, case_name: str) -> None:
        self.compose_file = str(E2E_DIR / case_name / "docker-compose.yml")
        self.project_name = f"chronos-e2e-{case_name}"

    def start(self) -> None:
        print(f"[infra] Starting {self.project_name}...")
        subprocess.run(
            ["docker", "compose", "-f", self.compose_file, "-p", self.project_name, "up", "-d", "--wait"],
            check=True,
            timeout=180,
        )
        print(f"[infra] {self.project_name} is up")

    def stop(self) -> None:
        print(f"[infra] Stopping {self.project_name}...")
        try:
            subprocess.run(
                ["docker", "compose", "-f", self.compose_file, "-p", self.project_name, "down", "-v", "--remove-orphans"],
                check=True,
                timeout=60,
            )
        except Exception as e:
            print(f"[infra] Warning during stop: {e}")
